File: 1114. Print in Order/solution.java
class Foo {
  private Semaphore firstCompleted;
  private Semaphore secondCompleted;

  public Foo() {
    firstCompleted = new Semaphore(0);
    secondCompleted = new Semaphore(0);
  }

  public void first(Runnable printFirst) throws InterruptedException {
    // printFirst.run() outputs "first". Do not change or remove this line.
    printFirst.run();
    firstCompleted.release();
  }

  public void second(Runnable printSecond) throws InterruptedException {
    firstCompleted.acquire();
    // printSecond.run() outputs "second". Do not change or remove this line.
    printSecond.run();
    secondCompleted.release();
  }

  public void third(Runnable printThird) throws InterruptedException {
    secondCompleted.acquire();
    // printThird.run() outputs "third". Do not change or remove this line.
    printThird.run();
  }
}

File: 1114. Print in Order/solution.py
import threading

class Foo(object):
  def __init__(self):
    self.firstCompleted = threading.Event()
    self.secondCompleted = threading.Event()

  def first(self, printFirst):
    """
    :type printFirst: method
    :rtype: void
    """
    # printFirst() outputs "first". Do not change or remove this line.
    printFirst()
    self.firstCompleted.set()

  def second(self, printSecond):
    """
    :type printSecond: method
    :rtype: void
    """
    self.firstCompleted.wait()
    # printSecond() outputs "second". Do not change or remove this line.
    printSecond()
    self.secondCompleted.set()
            
  def third(self, printThird):
    """
    :type printThird: method
    :rtype: void
    """
    self.secondCompleted.wait()
    # printThird() outputs "third". Do not change or remove this line.
    printThird()